### PR TITLE
[Property Wrappers] Refine assign_by_wrapper lowering

### PR DESCRIPTION
**Pull request by hborla**
If all of `self` is not yet initialized but the backing property wrapper is, lower `assign_by_wrapper` to re-assignment of the backing storage.

Resolves: [SR-13495](https://bugs.swift.org/browse/SR-13495)
Resolves: rdar://problem/68314811

### CODE REVIEW

File: test/SILOptimizer/di_property_wrappers.swift [hunk 1]
```diff
@@ -535,6 +535,54 @@ func testSR_12341() {
...
+
+  // CHECK-NEXT:   .. init TestLeak(42)
+  // CHECK-NEXT:     .. TestLeak(42).deinit
+  // CHECK-NEXT:   .. set TestLeak(27)
+  // CHECK-NEXT:     .. TestLeak(27).deinit
+  _ = SR_13495()
+
+  // CHECK-NEXT:   .. init TestLeak(42)
+  // CHECK-NEXT:     .. TestLeak(42).deinit
+  // CHECK-NEXT:   .. init TestLeak(27)
```
Comment (hborla) [inline test/SILOptimizer/di_property_wrappers.swift:576]: It seems wrong that this is doing initialization of the wrapper again, but I'm still trying to figure out how conditional "init or assign" is supposed to work

Comment (hborla) [reply test/SILOptimizer/di_property_wrappers.swift:576]: It looks me like `DIUseKind::InitOrAssign` always turns into unconditional initialization in `LifetimeChecker::handleConditionalInitAssign()`:

```
    // Finally, now that we know the value is uninitialized on all paths, it is
    // safe to do an unconditional initialization.
    Use.Kind = DIUseKind::Initialization;
```

Comment (hborla) [reply test/SILOptimizer/di_property_wrappers.swift:576]: Since this is an existing issue with property wrappers, I'll investigate separately

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

**Review by DougGregor — APPROVED** [review]
This looks good to me, but I'd also like the opinion of someone who has done more serious work in DI.

File: test/SILOptimizer/di_property_wrappers.swift [hunk 2]
```diff
@@ -535,6 +535,54 @@ func testSR_12341() {
   _ = SR_12341(condition: true)
 }
 
+class TestLeak: CustomStringConvertible {
```
Comment (slavapestov) [inline test/SILOptimizer/di_property_wrappers.swift:538]: Optional: you might want to stick a LifetimeTracked in here as well? StdlibUnittest will automatically assert if one leaks.

**Review by slavapestov — APPROVED** [review]

**Comment by hborla** [comment]
@swift-ci please smoke test

**Review by Yoshio123 — APPROVED** [review]

**Comment by hborla** [comment]
@swift-ci please smoke test

**Pull request merged**